### ISSUE TITLE
Problem in combination IBM CE and Lithops

### ISSUE THREAD

**Issue by sergii-mamedov** [lithops-cloud/lithops#1236]
Since last Saturday, we have been observing a strange situation in the combination of IBM Code Engine and Lithops. Approximately every 10-20 starts, I see that the IBM executor has completed its work
```
% ibmcloud ce jr list -s age
Listing job runs...
OK

Name                   Requested  Pending  Running  Failed  Succeeded  Unknown  Age    Mode  
lithops-cb59bf-4-m006  0          0        0        0       4          0        8m36s  task  
```
However, lithops logs are still waiting for a result from one of the executors:
```
2024-01-22 10:20:33,771 - DEBUG - lithops.monitor[Thread-31] - monitor.py:127 - ExecutorID cb59bf-4 - Pending: 1 - Running: 0 - Done: 3
2024-01-22 10:21:07,346 - DEBUG - lithops.monitor[Thread-31] - monitor.py:127 - ExecutorID cb59bf-4 - Pending: 1 - Running: 0 - Done: 3
```
@JosepSampe maybe you have an idea why this happened? How can I help debug this situation?

Additional info:
```
% ibmcloud ce jobrun events -n lithops-cb59bf-4-m006
Getting jobrun 'lithops-cb59bf-4-m006'...
Getting instances of jobrun 'lithops-cb59bf-4-m006'...
Getting events for all instances of job run 'lithops-cb59bf-4-m006'...
OK

lithops-cb59bf-4-m006-0-0:    
  Type    Reason     Age  Source                  Messages  
  Normal  Scheduled  23m  default-scheduler       Successfully assigned 5s4f5qcqf4f/lithops-cb59bf-4-m006-0-0 to 192.168.0.181  
  Normal  Pulling    23m  kubelet, 192.168.0.181  Pulling image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1"  
  Normal  Pulled     23m  kubelet, 192.168.0.181  Successfully pulled image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1" in 782.351965ms (782.365544ms including waiting)  
  Normal  Created    23m  kubelet, 192.168.0.181  Created container lithops-worker-301-69a73a4932  
  Normal  Started    23m  kubelet, 192.168.0.181  Started container lithops-worker-301-69a73a4932  

lithops-cb59bf-4-m006-1-0:    
  Type    Reason     Age  Source                  Messages  
  Normal  Scheduled  23m  default-scheduler       Successfully assigned 5s4f5qcqf4f/lithops-cb59bf-4-m006-1-0 to 192.168.4.251  
  Normal  Pulling    23m  kubelet, 192.168.4.251  Pulling image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1"  
  Normal  Pulled     23m  kubelet, 192.168.4.251  Successfully pulled image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1" in 770.523521ms (770.543344ms including waiting)  
  Normal  Created    23m  kubelet, 192.168.4.251  Created container lithops-worker-301-69a73a4932  
  Normal  Started    23m  kubelet, 192.168.4.251  Started container lithops-worker-301-69a73a4932  

lithops-cb59bf-4-m006-2-0:    
  Type    Reason     Age  Source                  Messages  
  Normal  Scheduled  23m  default-scheduler       Successfully assigned 5s4f5qcqf4f/lithops-cb59bf-4-m006-2-0 to 192.168.0.213  
  Normal  Pulling    23m  kubelet, 192.168.0.213  Pulling image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1"  
  Normal  Pulled     23m  kubelet, 192.168.0.213  Successfully pulled image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1" in 845.24095ms (845.261129ms including waiting)  
  Normal  Created    23m  kubelet, 192.168.0.213  Created container lithops-worker-301-69a73a4932  
  Normal  Started    23m  kubelet, 192.168.0.213  Started container lithops-worker-301-69a73a4932  

lithops-cb59bf-4-m006-3-0:    
  Type    Reason     Age  Source                 Messages  
  Normal  Scheduled  23m  default-scheduler      Successfully assigned 5s4f5qcqf4f/lithops-cb59bf-4-m006-3-0 to 192.168.2.17  
  Normal  Pulling    23m  kubelet, 192.168.2.17  Pulling image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1"  
  Normal  Pulled     23m  kubelet, 192.168.2.17  Successfully pulled image "docker.io/metaspace2020/metaspace-lithops-ce:3.0.1" in 791.733504ms (791.776368ms including waiting)  
  Normal  Created    23m  kubelet, 192.168.2.17  Created container lithops-worker-301-69a73a4932  
  Normal  Started    23m  kubelet, 192.168.2.17  Started container lithops-worker-301-69a73a4932  
```
and
```
% ibmcloud ce jr get -n lithops-cb59bf-4-m006
Getting jobrun 'lithops-cb59bf-4-m006'...
Getting instances of jobrun 'lithops-cb59bf-4-m006'...
Getting events of jobrun 'lithops-cb59bf-4-m006'...
For troubleshooting information visit: https://cloud.ibm.com/docs/codeengine?topic=codeengine-troubleshoot-job.
Run 'ibmcloud ce jobrun events -n lithops-cb59bf-4-m006' to get the system events of the job run instances.
Run 'ibmcloud ce jobrun logs -f -n lithops-cb59bf-4-m006' to follow the logs of the job run instances.
OK

Name:          lithops-cb59bf-4-m006  
ID:            e01bc972-149b-461b-8a7e-36318ab07100  
Project Name:  metaspace-prod-ce  
Project ID:    d8d40498-98b7-4105-bd8d-edb51f40225d  
Age:           23m  
Created:       2024-01-22T09:59:08+01:00  

Job Ref:                lithops-worker-301-69a73a4932  
Commands:                 
  /usr/local/bin/python  
Arguments:                
  /lithops/lithopsentry.py  
  $(ACTION)  
  $(PAYLOAD)  
Environment Variables:    
  Type                     Name     Value  
  Literal                  ACTION   run  
  ConfigMap key reference  PAYLOAD  lithops-cb59bf-4-m006.lithops.payload  
Image:                  docker.io/metaspace2020/metaspace-lithops-ce:3.0.1  
Resource Allocation:      
  CPU:                1  
  Ephemeral Storage:  400M  
  Memory:             2G  

Runtime:      
  Mode:                  task  
  Array Indices:         0-3  
  Array Size:            4  
  JOB_ARRAY_SIZE Value:  4  
  Max Execution Time:    1800  
  Retry Limit:           2  

Status:       
  Completed:          22m  
  Instance Statuses:    
    Succeeded:  4  
  Conditions:         
    Type      Status  Last Probe  Last Transition  Reason  
    Pending   True    23m         23m                
              True    23m         23m                
    Running   True    23m         23m                
    Complete  True    22m         22m                

Events:       
  Type    Reason     Age                 Source                Messages  
  Normal  Updated    23m (x10 over 23m)  batch-job-controller  Updated JobRun "lithops-cb59bf-4-m006"  
  Normal  Completed  23m                 batch-job-controller  JobRun completed successfully  

Instances:    
  Name                       Running  Status     Restarts  Age  
  lithops-cb59bf-4-m006-0-0  0/1      Succeeded  0         23m  
  lithops-cb59bf-4-m006-1-0  0/1      Succeeded  0         23m  
  lithops-cb59bf-4-m006-2-0  0/1      Succeeded  0         23m  
  lithops-cb59bf-4-m006-3-0  0/1      Succeeded  0         23m  

```

**Comment by JosepSampe** [comment]
Hi @sergii-mamedov, from the Code Engine perspective everything looks good. Is it still possible for you to access the Lithops logs inside the containers? Just want to see if some of them produced an error while storing the final status or result file in the object store, or the issue comes from a different part

**Comment by JosepSampe** [comment]
Yes, that's weird, it seems you can see the container `lithops-cb59bf-4-m006-0-0/lithops-worker-301-69a73a4932` and describe it, but there are no logs at all. I think It won't be easy to see if it is a Lithops issue without the worker logs. Might this be something related to an Out Of Memory issue?

**Comment by sergii-mamedov** [comment]
> Yes, that's weird, it seems you can see the container lithops-cb59bf-4-m006-0-0/lithops-worker-301-69a73a4932 and describe it, but there are no logs at all.

I do not know how to do it. Please provide instruction if you know how.

> Might this be something related to an Out Of Memory issue?

I think no, because we can handle with OOM. And now it is happens with a different datasets and different step.

**Comment by JosepSampe** [comment]
What I can see is that in your client logs, the function is in pending status:
```
2024-01-22 10:20:33,771 - DEBUG - lithops.monitor[Thread-31] - monitor.py:127 - ExecutorID cb59bf-4 - Pending: 1 - Running: 0 - Done: 3
```

All the functions write a 0 byte file in COS when they start, so that we know from the client at what moment they started and we can change the state from `pending` to `running`.

```
2024-01-22 08:59:11,887 [DEBUG] ibm_cos.py:148 -- PUT Object lithops.jobs/cb59bf-4-M006/00003/3ef4991b2d4e.init - Size: 0.0B - OK  
```
 In your case the function is always in `pending` status, so yes, it is not an OOM error, because this means your code didn't start executing. 

I have the feeling that most probably the lithops code didn't even start executing, because at least you should see a few log lines if it fails because of a Lithops issue somewhere. What I don't understand is why you see the instance as `Succeeded`:
```
Instances:    
  Name                       Running  Status     Restarts  Age  
  lithops-cb59bf-4-m006-0-0  0/1      Succeeded  0         23m 
```
Will continue investigating. Do you experience this issue all the time? or from time to time? Do you experience it in the same function execution? or with different functions?

**Comment by sergii-mamedov** [comment]
It started on Friday. We haven't changed our code in a few weeks. We have this behavior for different steps (functions), with different number of executors and different amount of allocated resources.

**Comment by sergii-mamedov** [comment]
@JosepSampe 
Yesterday, we made an emergency switch to AWS. Because of this, the combination of IBM CE and Lihtops is no longer relevant for us. But for AWS, I saw a similar problem for one dataset (when one of the executors does not return a result). I ran this dataset three times - the problem is reproduced. I will try to reproduce it on staging and enable additional logs. I hope this helps us understand the root of the problem.
